### PR TITLE
Reduce dependency on MatrixWorkspace in DiscusMultipleScatteringCorrection

### PR DESCRIPTION
**Pull request by DannyHindson**
**Description of work.**

These changes remove the use of the MatrixWorkspace class from the internal calculation logic in the DiscusMultipleScatteringCorrection algorithm and replaces it with some simple data structures for 2D data based on std::vector. The MatrixWorkspace class is now only used to read input data and save the results to the ADS at the end.

The aim of this was to more loosely couple the algorithm to Mantid so that it might be usable by a wider audience eg Scipp. Although note - there still is significant Mantid coupling (eg in the use of the Mantid Geometry framework) but thought it would be a useful exercise to see how difficult it would be to loosen the coupling in terms of the workspaces at least.

**To test:**

Main things to check are that the algorithm is functionally the same and that I've not slowed it down. The unit tests on the algorithm are pretty comprehensive so hopefully they will cover the "functionally the same" part. Re performance I suggest running this Discus calculation on a release build before\after these changes and confirm it doesn't run any slower with these changes. It should take about a minute although depends on the power of your machine. Test files here [TestFiles.zip](https://github.com/mantidproject/mantid/files/10001709/TestFiles.zip):
```
# import mantid algorithms, numpy and matplotlib

from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np
import math

folder = r"C:\Multiple Scattering\"

sqw = Load(Filename=os.path.join(folder,'irs17901_multi_graphite002_sqw_widerq.nxs'))

sqw = Rebin(sqw,'4,0.1,21')

inputWS = Load(Filename=os.path.join(folder,'irs17901_multi_graphite002_red.nxs'))

# default cylinder axis is "up" ie y
SetSample(InputWorkspace=inputWS,
          Geometry={'Shape': 'Cylinder', 'Height': 3.0,'Radius': 1.0,'Center': [0.,0.,0.]},
          Material={'ChemicalFormula': 'He'})

DiscusMultipleScatteringCorrection(InputWorkspace=inputWS, StructureFactorWorkspace=sqw,
                                   OutputWorkspace="MuscatResultsHe", NeutronPathsSingle=200,
                                   NeutronPathsMultiple=1000, NormalizeStructureFactors=False)
```

*There is no associated issue.*

*This does not require release notes* because **there is no functional change**

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp [hunk 1]
```diff
@@ -75,6 +75,26 @@ struct EFixedProvider {
...
 
+std::unique_ptr<DiscusData2D> DiscusData2D::CreateCopy(bool clearY) {
+  auto data2DNew = std::make_unique<DiscusData2D>();
+  data2DNew->m_data.resize(m_data.size());
+  for (size_t i = 0; i < m_data.size(); i++) {
+    data2DNew->m_data[i].X = m_data[i].X;
+    if (clearY)
+      data2DNew->m_data[i].Y = std::vector<double>(m_data[i].Y.size(), 0.);
+    else
+      data2DNew->m_data[i].Y = m_data[i].Y;
```
Comment (robertapplin) [inline Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:86]:
```suggestion
    data2DNew->m_data[i].Y = clearY ? std::vector<double>(m_data[i].Y.size(), 0.) : m_data[i].Y;
```
Just a suggestion, no need to change if you think its not as easy to read

File: Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp [hunk 2]
```diff
@@ -1807,11 +1835,13 @@ DiscusMultipleScatteringCorrection::createSparseWorkspace(const API::MatrixWorks
 void DiscusMultipleScatteringCorrection::createInvPOfQWorkspaces(ComponentWorkspaceMappings &matWSs, size_t nhists) {
   for (auto &SQWSMapping : matWSs) {
     auto &QSQ = SQWSMapping.QSQ;
-    size_t expectedMaxSize = QSQ->size();
-    MatrixWorkspace_sptr ws = DataObjects::create<Workspace2D>(nhists, HistogramData::Points{0.});
-    ws->dataX(0).reserve(expectedMaxSize);
+    size_t expectedMaxSize =
+        std::accumulate(QSQ->histograms().begin(), QSQ->histograms().end(), static_cast<size_t>(0),
```
Comment (robertapplin) [inline Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:1839]:
```suggestion
        std::accumulate(QSQ->histograms().cbegin(), QSQ->histograms().cend(), static_cast<size_t>(0),
```

File: Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp [hunk 3]
```diff
@@ -362,56 +382,70 @@ void DiscusMultipleScatteringCorrection::prepareStructureFactors() {
...
+  } else if (SQWS->getAxis(1)->isSpectra()) {
+    // for elastic set w=0 on the spectrum axis to align code with inelastic
+    auto newAxis = std::make_unique<NumericAxis>(std::vector<double>{0.});
+    newAxis->setUnit("DeltaE");
+    SQWS->replaceAxis(1, std::move(newAxis));
+  }
+  auto specAxis = dynamic_cast<NumericAxis *>(SQWS->getAxis(1));
+  std::vector<DiscusData1D> data;
+  for (size_t i = 0; i < SQWS->getNumberHistograms(); i++) {
+    data.push_back(DiscusData1D{SQWS->histogram(i).dataX(), SQWS->histogram(i).dataY()});
```
Comment (robertapplin) [inline Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:441]: Would it be better to use `emplace_back` here and just provide the arguments of the `DiscusData1D` constructor? I think that would then avoid a move

**Review by robertapplin — CHANGES_REQUESTED** [review]
I have a few code suggestions, some of them picky!

The run times looked good to me:
`this-branch` 42s
`main` 46s

File: Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp (hunk 2 above)
Comment (DannyHindson) [reply Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:1839]: Yes will change. I copied the code from Workspace2D::size() where it uses the non-const iterators for some reason

File: Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp (hunk 3 above)
Comment (DannyHindson) [reply Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp:441]: Good spot. I'll change to emplace_back. The compiler seems to want a constructor added to DiscusData1D that takes the 2 parameters but that's no big deal to add. I seem to have used push_back elsewhere in this unit as well. The only other instance that looks wrong is in DiscusMultipleScatteringCorrection::updateWeightAndPosition so I'll change that to emplace_back also

Comment (robertapplin) [reply Framework/Algorithms/src/DiscusMultipleScatteringCorrection.cpp]: Great, thank you!

**Review by robertapplin — APPROVED** [review]
Thanks for the changes. It looks good and works as expected

**Pull request merged**